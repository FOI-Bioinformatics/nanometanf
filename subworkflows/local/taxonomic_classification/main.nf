/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    SUBWORKFLOW: TAXONOMIC_CLASSIFICATION  
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Multi-tool taxonomic classification with taxpasta standardization
    
    Supported classifiers:
    - kraken2: Kraken2 k-mer based classification
    
    Future classifiers (ready to implement):
    - centrifuge: Centrifuge classification
    - metaphlan: MetaPhlAn4 marker-based profiling
    - kaiju: Kaiju protein-level classification
    
    Features:
    - Tool-agnostic interface
    - Taxpasta standardization for consistent outputs
    - Easy addition of new classifiers
----------------------------------------------------------------------------------------
*/

include { KRAKEN2_KRAKEN2                } from "${projectDir}/modules/nf-core/kraken2/kraken2/main"
include { KRAKEN2_OPTIMIZED              } from "${projectDir}/modules/local/kraken2_optimized/main"
// NOTE: Incremental classification feature disabled - modules not implemented
// include { KRAKEN2_INCREMENTAL_CLASSIFIER } from "${projectDir}/modules/local/kraken2_incremental_classifier/main"
// include { KRAKEN2_OUTPUT_MERGER          } from "${projectDir}/modules/local/kraken2_output_merger/main"
// include { KRAKEN2_REPORT_GENERATOR       } from "${projectDir}/modules/local/kraken2_report_generator/main"
include { TAXPASTA_STANDARDISE           } from "${projectDir}/modules/nf-core/taxpasta/standardise/main"

workflow TAXONOMIC_CLASSIFICATION {

    take:
    ch_reads     // channel: [ val(meta), path(reads) ]
    ch_db        // channel: [ path(db) ]

    main:
    ch_versions = Channel.empty()
    ch_classified_reads = Channel.empty()
    ch_unclassified_reads = Channel.empty()
    ch_reads_assignment = Channel.empty()
    ch_raw_reports = Channel.empty()

    // Set classifier and validate parameters
    def classifier = params.classifier ?: 'kraken2'
    def output_format = params.taxpasta_format ?: 'tsv'

    //
    // PHASE 2 OPTIMIZATION: Automatic database preloading in real-time mode
    //
    // Enable Kraken2 optimizations automatically for PromethION real-time processing
    // Memory-mapping enables OS-level database caching, eliminating redundant disk I/O across batches
    //
    def auto_enable_optimizations = params.realtime_mode && !params.kraken2_use_optimizations
    def use_memory_mapping = params.realtime_mode ? true : params.kraken2_memory_mapping
    def use_optimizations = params.realtime_mode ? true : params.kraken2_use_optimizations

    if (auto_enable_optimizations && classifier == 'kraken2') {
        log.info "=== Phase 2: Database Preloading Enabled ==="
        log.info "Real-time mode detected - automatically enabling Kraken2 optimizations:"
        log.info "  - Memory-mapped database loading: ENABLED"
        log.info "  - Database cached in OS page cache for reuse across batches"
        log.info "  - Eliminates repeated database loading (30+ batches â†’ 1 load)"
        log.info "  - Estimated time savings: 1-3 minutes per batch after first load"
    }
    
    //
    // BRANCH: Route to appropriate classifier
    //
    switch(classifier) {
        case 'kraken2':
            //
            // MODULE: Run Kraken2 for taxonomic classification
            // Two modes: optimized or standard (incremental disabled - modules not implemented)
            //
            if (false && params.kraken2_enable_incremental == true) {  // Incremental mode disabled
                log.info "Using incremental Kraken2 processing with batch caching"
                log.info "  - Raw outputs cached per batch for efficient merging"
                log.info "  - Cumulative reports generated from merged data"

                //
                // MODULE: Incremental classification with caching
                //
                KRAKEN2_INCREMENTAL_CLASSIFIER (
                    ch_reads,
                    ch_db,
                    params.save_output_fastqs ?: false,
                    params.save_reads_assignment ?: false,
                    params.kraken2_cache_dir ?: "${params.outdir}/cache/kraken2"
                )
                ch_versions = ch_versions.mix(KRAKEN2_INCREMENTAL_CLASSIFIER.out.versions)

                // Collect batch outputs per sample/barcode
                ch_raw_kraken2_outputs = KRAKEN2_INCREMENTAL_CLASSIFIER.out.raw_kraken2_output
                    .groupTuple(by: 0)  // Group by meta

                ch_batch_reports = KRAKEN2_INCREMENTAL_CLASSIFIER.out.report
                    .groupTuple(by: 0)  // Group by meta

                ch_batch_metadata = KRAKEN2_INCREMENTAL_CLASSIFIER.out.batch_metadata
                    .groupTuple(by: 0)  // Group by meta
                    .map{ meta, files -> files }  // Extract just the metadata files (remove meta)

                //
                // MODULE: Merge batch outputs
                //
                KRAKEN2_OUTPUT_MERGER (
                    ch_raw_kraken2_outputs,
                    ch_batch_metadata
                )
                ch_versions = ch_versions.mix(KRAKEN2_OUTPUT_MERGER.out.versions)

                //
                // MODULE: Generate cumulative report
                //
                KRAKEN2_REPORT_GENERATOR (
                    KRAKEN2_OUTPUT_MERGER.out.cumulative_output,
                    ch_batch_reports,
                    ch_db
                )
                ch_versions = ch_versions.mix(KRAKEN2_REPORT_GENERATOR.out.versions)

                // Collect outputs from incremental path
                ch_classified_reads = KRAKEN2_INCREMENTAL_CLASSIFIER.out.classified_reads_fastq
                ch_unclassified_reads = KRAKEN2_INCREMENTAL_CLASSIFIER.out.unclassified_reads_fastq
                ch_reads_assignment = KRAKEN2_INCREMENTAL_CLASSIFIER.out.classified_reads_assignment
                ch_raw_reports = KRAKEN2_REPORT_GENERATOR.out.report
                ch_performance_metrics = KRAKEN2_REPORT_GENERATOR.out.stats

            } else if (use_optimizations == true) {
                log.info "Using optimized Kraken2 classification with:"
                log.info "  - Memory mapping: ${use_memory_mapping}"
                log.info "  - Confidence threshold: ${params.kraken2_confidence}"
                log.info "  - Minimum hit groups: ${params.kraken2_minimum_hit_groups}"

                KRAKEN2_OPTIMIZED (
                    ch_reads,
                    ch_db,
                    params.save_output_fastqs ?: false,
                    params.save_reads_assignment ?: false,
                    use_memory_mapping,
                    params.kraken2_confidence ?: 0.0,
                    params.kraken2_minimum_hit_groups ?: 0
                )
                ch_versions = ch_versions.mix(KRAKEN2_OPTIMIZED.out.versions)

                // Collect outputs
                ch_classified_reads = KRAKEN2_OPTIMIZED.out.classified_reads_fastq
                ch_unclassified_reads = KRAKEN2_OPTIMIZED.out.unclassified_reads_fastq
                ch_reads_assignment = KRAKEN2_OPTIMIZED.out.classified_reads_assignment
                ch_raw_reports = KRAKEN2_OPTIMIZED.out.report
                ch_performance_metrics = KRAKEN2_OPTIMIZED.out.performance_metrics
            } else {
                KRAKEN2_KRAKEN2 (
                    ch_reads,
                    ch_db,
                    params.save_output_fastqs ?: false,    // save_output_fastqs
                    params.save_reads_assignment ?: false  // save_reads_assignment
                )
                ch_versions = ch_versions.mix(KRAKEN2_KRAKEN2.out.versions)

                // Collect outputs
                ch_classified_reads = KRAKEN2_KRAKEN2.out.classified_reads_fastq
                ch_unclassified_reads = KRAKEN2_KRAKEN2.out.unclassified_reads_fastq
                ch_reads_assignment = KRAKEN2_KRAKEN2.out.classified_reads_assignment
                ch_raw_reports = KRAKEN2_KRAKEN2.out.report
                ch_performance_metrics = Channel.empty()
            }
            break
            
        // Future classifiers to be added here:
        // case 'centrifuge':
        //     CENTRIFUGE_CENTRIFUGE(...)
        //     break
        // case 'metaphlan':
        //     METAPHLAN4_METAPHLAN4(...)
        //     break
        
        default:
            error "Unsupported classifier: ${classifier}. Currently supported: kraken2"
    }
    
    //
    // MODULE: Standardize classification reports with taxpasta
    //
    if (params.enable_taxpasta_standardization != false) {
        TAXPASTA_STANDARDISE (
            ch_raw_reports,
            classifier,
            output_format,
            params.taxonomy_file ? file(params.taxonomy_file, checkIfExists: true) : []
        )
        ch_versions = ch_versions.mix(TAXPASTA_STANDARDISE.out.versions)
        ch_standardized_reports = TAXPASTA_STANDARDISE.out.standardised_profile
    } else {
        ch_standardized_reports = ch_raw_reports
    }

    emit:
    classified_reads      = ch_classified_reads      // channel: [ val(meta), path(fastq) ]
    unclassified_reads    = ch_unclassified_reads    // channel: [ val(meta), path(fastq) ]
    reads_assignment      = ch_reads_assignment      // channel: [ val(meta), path(txt) ]
    report                = ch_raw_reports           // channel: [ val(meta), path(txt) ] - Original format for compatibility
    standardized_report   = ch_standardized_reports  // channel: [ val(meta), path(tsv/csv/etc) ] - Standardized format
    performance_metrics   = ch_performance_metrics   // channel: [ path(json) ] - Performance metrics (when optimizations enabled)
    classifier_used       = Channel.value(classifier) // channel: val(classifier_name)
    versions              = ch_versions              // channel: [ path(versions.yml) ]
}