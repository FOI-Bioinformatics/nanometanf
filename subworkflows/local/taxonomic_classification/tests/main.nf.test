nextflow_workflow {

    name "Test TAXONOMIC_CLASSIFICATION"
    script "../../taxonomic_classification.nf"
    workflow "TAXONOMIC_CLASSIFICATION"

    test("Should perform basic Kraken2 taxonomic classification") {

        setup {
            """
            # Create test data
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/classification_sample.fastq.gz
            
            # Create mock Kraken2 database directory
            mkdir -p $outputDir/mock_kraken2_db
            echo "Mock Kraken2 database" > $outputDir/mock_kraken2_db/hash.k2d
            echo "Mock taxonomy" > $outputDir/mock_kraken2_db/taxo.k2d
            echo "Mock options" > $outputDir/mock_kraken2_db/opts.k2d
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'classification_test', single_end: true],
                        file('$outputDir/classification_sample.fastq.gz')
                    ]
                ]
                input[1] = file('$outputDir/mock_kraken2_db')
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'tsv'
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.classified_reads
            assert workflow.out.reports
            assert workflow.out.versions
            
            // Verify Kraken2 processing
            def reports = workflow.out.reports.toList()
            assert reports.size() >= 1
            
            // Verify classified reads output
            def classified_reads = workflow.out.classified_reads.toList()
            assert classified_reads.size() >= 1
        }
    }

    test("Should handle multiple classification formats") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/format_sample.fastq.gz
            
            # Create mock database
            mkdir -p $outputDir/format_kraken2_db
            echo "Mock database for format testing" > $outputDir/format_kraken2_db/hash.k2d
            echo "Mock taxonomy data" > $outputDir/format_kraken2_db/taxo.k2d
            echo "Mock options file" > $outputDir/format_kraken2_db/opts.k2d
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'format_test', single_end: true],
                        file('$outputDir/format_sample.fastq.gz')
                    ]
                ]
                input[1] = file('$outputDir/format_kraken2_db')
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'biom'
                kraken2_save_output_fastqs = true
                kraken2_save_readclassifications = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.classified_reads
            assert workflow.out.unclassified_reads
            assert workflow.out.reports
            assert workflow.out.reads_assignment
            
            // Verify multiple output types
            def classified_reads = workflow.out.classified_reads.toList()
            assert classified_reads.size() >= 1
            
            def unclassified_reads = workflow.out.unclassified_reads.toList()
            assert unclassified_reads.size() >= 1
            
            def assignments = workflow.out.reads_assignment.toList()
            assert assignments.size() >= 1
        }
    }

    test("Should process multiple samples in batch") {

        setup {
            """
            # Create multiple test samples
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/batch_sample1.fastq.gz
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/batch_sample2.fastq.gz
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/batch_sample3.fastq.gz
            
            # Create mock database
            mkdir -p $outputDir/batch_kraken2_db
            echo "Mock batch database" > $outputDir/batch_kraken2_db/hash.k2d
            echo "Mock batch taxonomy" > $outputDir/batch_kraken2_db/taxo.k2d
            echo "Mock batch options" > $outputDir/batch_kraken2_db/opts.k2d
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'batch_sample1', single_end: true],
                        file('$outputDir/batch_sample1.fastq.gz')
                    ],
                    [
                        [id: 'batch_sample2', single_end: true],
                        file('$outputDir/batch_sample2.fastq.gz')
                    ],
                    [
                        [id: 'batch_sample3', single_end: true],
                        file('$outputDir/batch_sample3.fastq.gz')
                    ]
                ]
                input[1] = file('$outputDir/batch_kraken2_db')
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'tsv'
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '3.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.classified_reads
            assert workflow.out.reports
            
            // Verify batch processing
            def reports = workflow.out.reports.toList()
            assert reports.size() >= 3
            
            def classified_reads = workflow.out.classified_reads.toList()
            assert classified_reads.size() >= 3
        }
    }

    test("Should handle classification confidence thresholds") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/confidence_sample.fastq.gz
            
            # Create mock database
            mkdir -p $outputDir/confidence_kraken2_db
            echo "Mock confidence database" > $outputDir/confidence_kraken2_db/hash.k2d
            echo "Mock confidence taxonomy" > $outputDir/confidence_kraken2_db/taxo.k2d
            echo "Mock confidence options" > $outputDir/confidence_kraken2_db/opts.k2d
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'confidence_test', single_end: true],
                        file('$outputDir/confidence_sample.fastq.gz')
                    ]
                ]
                input[1] = file('$outputDir/confidence_kraken2_db')
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'tsv'
                kraken2_confidence = 0.75
                kraken2_minimum_base_quality = 20
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.classified_reads
            assert workflow.out.reports
            
            // Verify confidence-based processing
            def reports = workflow.out.reports.toList()
            assert reports.size() >= 1
        }
    }

    test("Should standardize output with Taxpasta") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/taxpasta_sample.fastq.gz
            
            # Create mock database
            mkdir -p $outputDir/taxpasta_kraken2_db
            echo "Mock taxpasta database" > $outputDir/taxpasta_kraken2_db/hash.k2d
            echo "Mock taxpasta taxonomy" > $outputDir/taxpasta_kraken2_db/taxo.k2d
            echo "Mock taxpasta options" > $outputDir/taxpasta_kraken2_db/opts.k2d
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'taxpasta_test', single_end: true],
                        file('$outputDir/taxpasta_sample.fastq.gz')
                    ]
                ]
                input[1] = file('$outputDir/taxpasta_kraken2_db')
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'biom'
                taxpasta_add_name = true
                taxpasta_add_rank = true
                taxpasta_add_lineage = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.reports
            assert workflow.out.standardised_reports
            
            // Verify Taxpasta standardization
            if (workflow.out.standardised_reports) {
                def standardised = workflow.out.standardised_reports.toList()
                assert standardised.size() >= 1
            }
        }
    }

    test("Should handle empty input gracefully") {

        when {
            workflow {
                """
                input[0] = []
                input[1] = file('$projectDir/assets/test_data')  // Use any existing directory
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'tsv'
                max_cpus = 1
                max_memory = '2.GB'
                max_time = '1.min'
            }
        }

        then {
            // Should handle empty input gracefully
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            
            if (workflow.success) {
                // Empty outputs expected
                if (workflow.out.classified_reads) {
                    def classified_reads = workflow.out.classified_reads.toList()
                    assert classified_reads.size() == 0
                }
            }
        }
    }

    test("Should support different Kraken2 database types") {

        setup {
            """
            cp $projectDir/tests/test_sample.fastq.gz $outputDir/db_type_sample.fastq.gz
            
            # Create mock standard database
            mkdir -p $outputDir/standard_kraken2_db
            echo "Standard mock database" > $outputDir/standard_kraken2_db/hash.k2d
            echo "Standard mock taxonomy" > $outputDir/standard_kraken2_db/taxo.k2d
            echo "Standard mock options" > $outputDir/standard_kraken2_db/opts.k2d
            """
        }

        when {
            workflow {
                """
                input[0] = [
                    [
                        [id: 'db_type_test', single_end: true],
                        file('$outputDir/db_type_sample.fastq.gz')
                    ]
                ]
                input[1] = file('$outputDir/standard_kraken2_db')
                """
            }
            
            params {
                classifier = 'kraken2'
                taxpasta_format = 'tsv'
                kraken2_use_names = true
                kraken2_report_zero_counts = true
                max_cpus = 2
                max_memory = '4.GB'
                max_time = '2.min'
            }
        }

        then {
            assert workflow.success
            assert snapshot(workflow.out.versions).match()
            assert workflow.out.classified_reads
            assert workflow.out.reports
            
            // Verify database-specific processing
            def reports = workflow.out.reports.toList()
            assert reports.size() >= 1
        }
    }
}