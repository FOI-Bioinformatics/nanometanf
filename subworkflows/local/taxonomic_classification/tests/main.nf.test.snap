{
    "Should handle multiple classification formats": {
        "content": [
            [
                "versions.yml:md5,47cea2c8122ce1997cb90d2dfcb64f57",
                "versions.yml:md5,b4e713197270631d10fc1b5c0697869c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:22:30.137745"
    },
    "Should perform basic Kraken2 taxonomic classification": {
        "content": [
            [
                "versions.yml:md5,47cea2c8122ce1997cb90d2dfcb64f57",
                "versions.yml:md5,b4e713197270631d10fc1b5c0697869c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:20:17.787946"
    },
    "Should support different Kraken2 database types": {
        "content": [
            [
                "versions.yml:md5,47cea2c8122ce1997cb90d2dfcb64f57",
                "versions.yml:md5,b4e713197270631d10fc1b5c0697869c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:23:06.073322"
    },
    "Should handle classification confidence thresholds": {
        "content": [
            [
                "versions.yml:md5,47cea2c8122ce1997cb90d2dfcb64f57",
                "versions.yml:md5,b4e713197270631d10fc1b5c0697869c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:22:43.972488"
    },
    "Should standardize output with Taxpasta": {
        "content": [
            [
                "versions.yml:md5,47cea2c8122ce1997cb90d2dfcb64f57",
                "versions.yml:md5,b4e713197270631d10fc1b5c0697869c"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-15T16:22:52.528962"
    }
}